### PR TITLE
Add Longest Increasing Subsequence

### LINKED ISSUE #153: Longest Increasing Subsequence
I would like to do an implementation of Longest Increasing Subsequence in C++.
Problem statement:
Given an unsorted array of integers, return the longest increasing subsequence.

### PR DESCRIPTION
**Pull request by shikherjaiswal**
Added Longest Increasing Subsequence code

- [x] Read the [contribution guidelines][contrib-guidelines]
- [x] Added [unit tests][unit-tests] (or unit tests have already been added)
- [ ] Updated the [Contents][contents]

Fixes #153 

[contrib-guidelines]: https://github.com/faheel/Algos/blob/master/CONTRIBUTING.md
[unit-tests]: https://github.com/faheel/Algos/blob/master/C%2B%2B/UNIT_TESTS.md
[contents]: https://github.com/faheel/Algos/tree/master/C%2B%2B#contents

### CODE REVIEW

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+
+/*
+  Longest Increasing Subsequence
+  ------------------------------
+  Given an unsorted array of integers, return the longest increasing subsequence length
+  Time complexity
```
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp]: Add newlines after each of these headers.

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  ---------------
+  O(N), where N is size of array
+  Author
+  ------
+  Shikhar Jaiswal(@shikherjaiswal)
+*/
+#ifndef longest_increasing_subsequence_hpp
+#define longest_increasing_subsequence_hpp
+#include <vector>
+int lis( std :: vector< int > arr, int n ) {
```
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp]: Add a newline and function description above function. Look at codebase for formatting tips.

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    int *lis, i, j, max = 0;
+    lis = (int*) malloc ( sizeof( int ) * n );
```
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp]: Use `new` instead of `malloc` and `delete` instead of `free`. Explanation: https://stackoverflow.com/questions/184537/in-what-cases-do-i-use-malloc-vs-new

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    /* Compute optimized LIS values in bottom up manner */
+    for (i = 1; i < n; i++ ){
+        for (j = 0; j < i; j++ )  {
+            if ( arr[i] > arr[j] && lis[i] < lis[j] + 1) {
+                lis[i] = lis[j] + 1;
+            }
+        }
+    }
+    /* Pick maximum of all LIS values */
+    for (i = 0; i < n; i++ ) {
```
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp]: Fix trailing whitespace after increment operators.

File: C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#include "third_party/catch.hpp"
+#include "algorithm/dynamic_programming/longest_increasing_subsequence.hpp"
+TEST_CASE("Base Cases", "[LIS]") {
```
Comment (alxmjo) [inline C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp:4]: Newline above.

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#include <vector>
+int lis( std :: vector< int > arr, int n ) {
+    int *lis, i, j, max = 0;
+    lis = (int*) malloc ( sizeof( int ) * n );
+    /* Initialize LIS values for all indexes */
+    for (i = 0; i < n; i++ ) {
+        lis[i] = 1;
+    }
+    /* Compute optimized LIS values in bottom up manner */
+    for (i = 1; i < n; i++ ){
```
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp]: Add space between ) and {, as in: 

```
for (...) {

}
```

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  ------
+  Shikhar Jaiswal(@shikherjaiswal)
+*/
+#ifndef longest_increasing_subsequence_hpp
+#define longest_increasing_subsequence_hpp
+#include <vector>
+int lis( std :: vector< int > arr, int n ) {
+    int *lis, i, j, max = 0;
+    lis = (int*) malloc ( sizeof( int ) * n );
+    /* Initialize LIS values for all indexes */
```
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp]: No need to block comment a single line comment, Just use `//`.

File: C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+  REQUIRE(lis(std :: vector < int > ({10, 22, 9, 33, 21, 50, 41, 60}),8) == 5);
+}
+TEST_CASE("Simple Cases", "[LIS]") {
```
Comment (alxmjo) [inline C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp:7]: Can you add a couple more test cases?

**Review by alxmjo — CHANGES_REQUESTED** [review]
Thanks for contributing! Here are a few notes.

**Comment by shikherjaiswal** [comment]
Added the requested changes and added more test cases @alxmjo

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+  Author
+  ------
+  Shikhar Jaiswal(@shikherjaiswal)
+*/
+#ifndef longest_increasing_subsequence_hpp
+#define longest_increasing_subsequence_hpp
+#include <vector>
+
+// lis function return longest increasing subsequence length given a vector and its size 'n'
```
Comment (alxmjo) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp]: This is more what I was thinking in terms of function comments: https://github.com/faheel/Algos/blob/reorganise/C%2B%2B/include/algorithm/backtracking/n_queens.hpp

**Comment by shikherjaiswal** [comment]
Updated the function comments in previous commit

**Comment by shikherjaiswal** [comment]
Please review @alxmjo

**Comment by alxmjo** [comment]
I get emails when you update, so no need to notify me again. I'll review when I have a moment.

**Review by alxmjo — APPROVED** [review]

**Comment by faheel** [comment]
@shikherjaiswal Please enter an entry in the contents (in `C++/README.md`) and fix the conflicts (by first pulling the latest changes from `reorganise`. Then your PR will be good to merge.

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+#ifndef longest_increasing_subsequence_hpp
+#define longest_increasing_subsequence_hpp
+
+#include <vector>
+
+/*
+    ==========================================================================
+    Public interface
+    ==========================================================================
```
Comment (faheel) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp:27]: No need of this comment.

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+*/
+
+
+/*
+    lis
+    ----------------
+    Returns the length of longest increasing subsequence of vector -arr
+*/
+
+int lis(std :: vector< int > arr, int n) {
```
Comment (faheel) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp:37]: * Rename the `lis` function to `longest_increasing_subsequence`.
* Remove unnecessary whitespace around `::`, `<`, `>` (please also make this change in other files you've added).
* What is `n` for? If it's value is always supposed to be the length of `arr` then it is not required. Otherwise, rename it to something more meaningful (like `max_index` or something).

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    ==========================================================================
+    Public interface
+    ==========================================================================
+*/
+
+
+/*
+    lis
+    ----------------
+    Returns the length of longest increasing subsequence of vector -arr
```
Comment (faheel) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp:34]: Update this comment to:
```
Returns the length of the longest increasing subsequence from the given vector of integers.
```

File: C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp [hunk 13]
```diff
@@ -0,0 +1,61 @@
+/*
+      Longest Increasing Subsequence
+      ------------------------------
+      Given an unsorted array of integers, return the longest increasing subsequence length
```
Comment (faheel) [inline C++/include/algorithm/dynamic_programming/longest_increasing_subsequence.hpp:4]: `[...] return the length of the longest increasing subsequence.`

File: C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp [hunk 14]
```diff
@@ -0,0 +1,13 @@
+#include "third_party/catch.hpp"
+#include "algorithm/dynamic_programming/longest_increasing_subsequence.hpp"
+
+TEST_CASE("Base Cases", "[LIS]") {
+  REQUIRE(lis(std :: vector < int > ({10, 22, 9, 33, 21, 50, 41, 60}),8) == 5);
+}
+TEST_CASE("Simple Cases", "[LIS]") {
```
Comment (faheel) [inline C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp:7]: "General cases" or "Normal cases" would be a better description.

File: C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp [hunk 15]
```diff
@@ -0,0 +1,13 @@
+#include "third_party/catch.hpp"
+#include "algorithm/dynamic_programming/longest_increasing_subsequence.hpp"
+
+TEST_CASE("Base Cases", "[LIS]") {
+  REQUIRE(lis(std :: vector < int > ({10, 22, 9, 33, 21, 50, 41, 60}),8) == 5);
```
Comment (faheel) [inline C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp:5]: This isn't a base case. Please move this under "General cases" and add the following cases here:
* `{}` (empty vector)
* `{0}`
* `{1}`
* `{-1}`
* `{0, 1}`
* `{1, 0}`
* `{-1, 0}`
* `{0, -1}`
* `{-1, 0, 1}` and a few other permutations of it.

File: C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+}
+TEST_CASE("Simple Cases", "[LIS]") {
+  REQUIRE(lis(std :: vector < int > ({2, 5, 3, 7, 11, 8, 10, 13, 6}),9) == 6);
+  REQUIRE(lis(std :: vector < int > ({58, 10, 45, -1, 100}),5) == 3);
+  REQUIRE(lis(std :: vector < int > ({-5, 22, 9, 33, 21, 1, 41, 60 }),8) == 5);
+  REQUIRE(lis(std :: vector < int > ({-5, 22, 9, -5, 21, 1, -8, 60 }),8) == 4);
+  REQUIRE(lis(std :: vector < int > ({-5, 21, 1, -8, 60}),5) == 3);
+}
```
Comment (faheel) [inline C++/test/algorithm/dynamic_programming/longest_increasing_subsequence.cpp:13]: It would be great if you can add some (just 2 or 3) larger cases too.

**Review by faheel — CHANGES_REQUESTED** [review]
Please enter an entry in the contents (in `C++/README.md`) and fix the conflicts (by first pulling the latest changes from `reorganise`.
